File: arduinobot_controller/arduinobot_controller/slider_control.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint



class SliderControl(Node):

    def __init__(self):
        super().__init__("slider_control")
      
        self.torso_pub_ = self.create_publisher(JointTrajectory, "torso_controller/joint_trajectory", 10)
        self.left_arm_pub_ = self.create_publisher(JointTrajectory, "left_arm_controller/joint_trajectory", 10)
        self.left_arm_gripper_pub_ = self.create_publisher(JointTrajectory, "left_arm_gripper_controller/joint_trajectory", 10)
        self.right_arm_pub_ = self.create_publisher(JointTrajectory, "right_arm_controller/joint_trajectory", 10)
        self.right_arm_gripper_pub_ = self.create_publisher(JointTrajectory, "right_arm_gripper_controller/joint_trajectory", 10)

        self.sub_ = self.create_subscription(JointState, "joint_commands", self.sliderCallback, 10)
        self.get_logger().info("Slider Control Node started")

    def sliderCallback(self, msg):
        torso_controller = JointTrajectory()
        left_arm_controller = JointTrajectory()
        left_arm_gripper_controller = JointTrajectory()
        right_arm_controller = JointTrajectory()
        right_arm_gripper_controller = JointTrajectory()
    
        torso_controller.joint_names = ["torso_joint"]
        left_arm_controller.joint_names = ["left_joint_1", "left_joint_2", "left_joint_3", "left_joint_4", "left_joint_5", "left_joint_6"]
        left_arm_gripper_controller.joint_names = ["left_gripper_left_claw_joint"]
        right_arm_controller.joint_names = ["right_joint_1", "right_joint_2", "right_joint_3", "right_joint_4", "right_joint_5", "right_joint_6"]
        right_arm_gripper_controller.joint_names = ["right_gripper_left_claw_joint"]

        torso_goal = JointTrajectoryPoint()
        left_arm_goal = JointTrajectoryPoint()
        left_arm_gripper_goal = JointTrajectoryPoint()
        right_arm_goal = JointTrajectoryPoint()
        right_arm_gripper_goal = JointTrajectoryPoint()

     

        torso_goal.positions = [msg.position[0]]
        left_arm_goal.positions = [msg.position[1:6]]
        left_arm_gripper_goal.positions = [msg.position[7]]
        right_arm_goal.positions = [msg.position[8:13]]
        right_arm_gripper_goal.positions = [msg.position[14]]

        torso_controller.points.append(torso_goal)
        left_arm_controller.points.append(left_arm_goal)
        left_arm_gripper_controller.points.append(left_arm_gripper_goal)
        right_arm_controller.points.append(right_arm_goal)
        right_arm_gripper_controller.points.append(right_arm_gripper_goal)

        self.torso_pub_.publish(torso_controller)
        self.left_arm_pub_.publish(left_arm_controller)
        self.left_arm_gripper_pub_.publish(left_arm_gripper_controller)
        self.right_arm_pub_.publish(right_arm_controller)
        self.right_arm_gripper_pub_.publish(right_arm_gripper_controller)


def main():
    rclpy.init()

    simple_publisher = SliderControl()
    rclpy.spin(simple_publisher)
    
    simple_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_controller/src/slider_control.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <trajectory_msgs/msg/joint_trajectory.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>

#include <chrono>

using namespace std::chrono_literals;
using std::placeholders::_1;

class SliderControl : public rclcpp::Node
{
public:
  SliderControl() : Node("slider_control")
  {
    sub_ = create_subscription<sensor_msgs::msg::JointState>(
      "joint_commands", 10, std::bind(&SliderControl::sliderCallback, this, _1));

    torso_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "torso_controller/joint_trajectory", 10);

    left_arm_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "left_arm_controller/joint_trajectory", 10);

    left_arm_gripper_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "left_arm_gripper_controller/joint_trajectory", 10);  

    right_arm_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "right_arm_controller/joint_trajectory", 10);

    right_arm_gripper_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>(
      "right_arm_gripper_controller/joint_trajectory", 10);  
    
    RCLCPP_INFO(get_logger(), "Slider Control Node started");
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr sub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr torso_pub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr left_arm_pub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr left_arm_gripper_pub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr right_arm_pub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr right_arm_gripper_pub_;

  void sliderCallback(const sensor_msgs::msg::JointState::SharedPtr msg) const
  {
    // Prepare trajectory messages for torso, left arm, and right arm
    trajectory_msgs::msg::JointTrajectory torso_command, left_arm_command, left_arm_gripper_command, right_arm_command, right_arm_gripper_command;
    torso_command.joint_names = {"torso_joint"};
    left_arm_command.joint_names = {"left_joint_1", "left_joint_2", "left_joint_3", "left_joint_4", "left_joint_5", "left_joint_6"};
    left_arm_gripper_command.joint_names = {"left_gripper_left_claw_joint"};
    right_arm_command.joint_names = {"right_joint_1", "right_joint_2", "right_joint_3", "right_joint_4", "right_joint_5", "right_joint_6"};
    right_arm_gripper_command.joint_names = {"right_gripper_left_claw_joint"};

    // Create joint trajectory points for each command
    trajectory_msgs::msg::JointTrajectoryPoint torso_goal, left_arm_goal, left_arm_gripper_goal, right_arm_goal, right_arm_gripper_goal;

    
    if (msg->position.size() >= 15) {
        torso_goal.positions.insert(torso_goal.positions.end(), msg->position.begin(), msg->position.begin() + 1);  // Only first position for torso
        for (size_t i = 1; i <= 6; ++i) {
            left_arm_goal.positions.push_back(msg->position[i]);  // Positions for the left arm
        }

        left_arm_gripper_goal.positions.push_back(msg->position[7]);

        for (size_t i = 8; i < 13; ++i) {
            right_arm_goal.positions.push_back(msg->position[i]);  // Positions for the right arm
        }

        right_arm_gripper_goal.positions.push_back(msg->position[14]);

       
        torso_goal.time_from_start = rclcpp::Duration(1s);
        left_arm_goal.time_from_start = rclcpp::Duration(1s);
        left_arm_gripper_goal.time_from_start = rclcpp::Duration(1s);
        right_arm_goal.time_from_start = rclcpp::Duration(1s);
        right_arm_gripper_goal.time_from_start = rclcpp::Duration(1s);

        // Add goals to commands
        torso_command.points.push_back(torso_goal);
        left_arm_command.points.push_back(left_arm_goal);
        left_arm_gripper_command.points.push_back(left_arm_gripper_goal);
        right_arm_command.points.push_back(right_arm_goal);
        right_arm_gripper_command.points.push_back(right_arm_gripper_goal);

        // Publish the commands
        torso_pub_->publish(torso_command);
        left_arm_pub_->publish(left_arm_command);
        left_arm_gripper_pub_->publish(left_arm_gripper_command);
        right_arm_pub_->publish(right_arm_command);
        right_arm_gripper_pub_->publish(right_arm_gripper_command);

    } else {
        RCLCPP_WARN(get_logger(), "Insufficient joint positions received.");
        return; 
    }
    
    
    
  }
};

int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SliderControl>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
